simcon_project/users/views/create_assignment.py

```diff
@@ -0,0 +1,156 @@
+    researcher=request.user
+    #researcher='researcher@111.com'
+    students=data.get('stuData')
+    templates=data.get('tempData')
+    labels=data.get('labelData')
+
+    #Transfer string type list to list type
+    students=decode(students)
+    templates=decode(templates)
+    labels=decode(labels)
+
+    #Verify date
+    today=datetime.datetime.combine(datetime.date.today(), datetime.time(0, 0))
+    dateTmp=datetime.datetime.strptime(date, "%Y-%m-%d")
+    assignToday=False
+    if dateTmp==today:
+        assignToday=True
+    if dateTmp<today:
+        success=1
+        errMsg=errMsg+'The assigned date cannot before today.\n\n'
+
+    #Save the data that can be directly assigned to assignment.
+    assignment=Assignment()
+    assignment.name=name
+    assignment.date_assigned=date
+    researcher=Researcher.objects.get(email=researcher)
+    assignment.researcher=researcher
+    assignment.save()
+
+    #Determine if students, labels, and templates are empty.
+    stuIsNull=isNull(students)
+    labelIsNull=isNull(labels)
+    tempIsNull=isNull(templates)
+
+    #Assign student information to assignment,
+    #and judge whether there is an error that both students and labels are empty.
+    if stuIsNull:
+        if labelIsNull:
+            success=1
+            errMsg=errMsg+'One of the students or labels must not be empty.\n\n'
+    else:
+        for stu in students:
+            stuTmp=Student.objects.get(email=stu)
+            assignment.students.add(stuTmp)
+
+    #Assign template information to assignment
+    if tempIsNull:
+        success=1
+        errMsg=errMsg+'Template must not be empty.\n\n'
+    else:
+        for temp in templates:
+            tempList=temp.split('--')
+            tempName=tempList[0]
+            tempDate=tempList[1]
+            tempTmp=ConversationTemplate.objects.get(name=tempName,creation_date=tempDate)
+            assignment.conversation_templates.add(tempTmp)
+
+    #Assign label information to assignment
+    if not labelIsNull:
+        for label in labels:
+            labelTmp=SubjectLabel.objects.get(label_name=label)
+            assignment.subject_labels.add(labelTmp)
+
+    #Send email
+    job_defaults = { 'max_instances': 20 }
+    tz = get_localzone()
+    sched = BackgroundScheduler(timezone=tz, job_defaults=job_defaults)
+
+    subject='Simulated Conversation Assignment Update'
+    msg='You received this email because you have a new assignment: '+name+'. Please check the assignment page.'
+    schedId='Assignment--'+name+'--'+date
+    #when an error occurs, there is no need to add this task to the schedule.
+    researcher='simulated.conversation@mail.com'
+    if success==0:
+        #If the assignment date is the same day.
+        if assignToday:
+            sendMail(subject, msg, students, researcher)
+        else:
+            #To test the sending mail function, change the content of run_date to the time you want.
+            #run_date='2021-01-22 18:35:00'
```

Comment:
Do you get to choose the time?

Reply:
> Do you get to choose the time?

I am not quite sure what you mean. We can change the time of sending the email by changing the run_date in the code. This is a function that may be used during testing. Because the specific time of assignment cannot be set from the researcher's perspective, only the assigned date can be selected. By default, the mail will be sent at midnight, which is not convenient for testing. Changing the run_date in the code can set the time of sending the email to two minutes later.